sklearn/exceptions.py

```diff
@@ -0,0 +1,125 @@
+
+
+class ConvergenceWarning(UserWarning):
+    """Custom warning to capture convergence problems"""
+
+
+class DataConversionWarning(UserWarning):
+    """Warning used to notify implicit data conversions happening in the code.
+
+    This warning occurs when some input data needs to be converted or
+    interpreted in a way that may not match the user's expectations.
+
+    For example, this warning may occur when the the user
+        - passes an integer array to a function which is essentially
+          real-valued;
+        - requests a non-copying operation, but a copy is required to meet the
+          implementation's data-type expectations;
+        - passes and input whose shape can be be interpreted ambiguously.
+    """
+
+
+class DataDimensionalityWarning(UserWarning):
+    """Custom warning to notify potential issues with data dimensionality.
+
+    For example, in random projection, this warning is raised when the
+    number of components, which quantifes the dimensionality of the target
+    projection space, is higher than the number of features, which quantifies
+    the dimensionality of the original source space, to imply that the
+    dimensionality of the problem will not be reduced.
+    """
+
+
+class EfficiencyWarning(UserWarning):
+    """Warning used to notify the user that the efficiency is affected.
+
+    This warning notifies the user that the efficiency may not be optimal due
+    to some reason which may be included as a part of the warning message.
+    This may be subclassed into a more specific Warning class.
+    """
+
+
+class FitFailedWarning(RuntimeWarning):
+    """Warning class used if there is an error while fitting the estimator.
+
+    This Warning is primarily used with meta estimators like GridSearchCV,
+    RandomizedSearchCV etc. to warn when there is an error while fitting the
+    estimator.
+
+    Examples
+    --------
+    >>> from sklearn.grid_search import GridSearchCV
+    >>> from sklearn.svm import LinearSVC
+    >>> from sklearn.exceptions import FitFailedWarning
+    >>> import warnings
+    >>> warnings.simplefilter('always', FitFailedWarning)
+    >>> gs = GridSearchCV(LinearSVC(), {'C': [-1, -2]}, error_score=0)
+    >>> X, y = [[1, 2], [3, 4], [5, 6], [7, 8], [8, 9]], [0, 0, 0, 1, 1]
+    >>> with warnings.catch_warnings(record=True) as w:
+    ...     try:
+    ...         gs.fit(X, y)   # This will raise a ValueError since C is < 0
+    ...     except ValueError:
+    ...         pass
+    ...     print("The warning raised was : %r" % w[-1].message)
+    ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    The warning raised was : 'FitFailedWarning("Classifier fit failed. The
+    score on this train-test partition for these parameters will be set to
+    0.000000. Details: \\\\nValueError(\\'Penalty term must be positive; got
+    (C=-2)\\',)",)'
+    """
+
+
+class NeighborsWarning(UserWarning):
+    """"""
+    pass
```

Comment:
This warning class is no longer used in scikit-learn. As it was probably never meant to be part of the public API I would be +1 for just removing it.

If people think that removing it is a backward compat breakage, maybe we should instead make it explicit in the docstring that this warning class is deprecated in 0.17 and will be removed in 0.19.

I don't see a better way to mark the deprecation of a warning class.